include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h

```diff
@@ -154,17 +274,201 @@ class PreconditionerML : public PreconditionerBase<double>
     // re-calculate matrix
     pde_operator.calculate_system_matrix(system_matrix);
 
-    // initialize Trilinos' AMG
-    amg.initialize(system_matrix, ml_data);
+    if(operator_type == AMGOperatorType::Default)
+    {
+      // Default setup.
+      amg.initialize(system_matrix, ml_data);
+    }
+    else
+    {
+      // get Teuchos::ParameterList to provide custom near null space basis
+      Teuchos::ParameterList parameter_list = get_parameter_list();
+
+      std::vector<VectorType> constant_modes = get_constant_modes();
+
+      // Add constant modes to Teuchos::ParameterList.
+      std::unique_ptr<Epetra_MultiVector>
+        ptr_distributed_modes; // has to stay alive until after amg.initialize();
+      set_operator_nullspace(parameter_list,
+                             ptr_distributed_modes,
+                             system_matrix.trilinos_matrix(),
+                             constant_modes);
+
+      // Initialize with the Teuchos::ParameterList.
+      amg.initialize(system_matrix, parameter_list);
+    }
 
     this->update_needed = false;
   }
 
 private:
+  Teuchos::ParameterList
+  get_parameter_list()
```

Comment:
I think this function should be realized as a free function with `MLData` as a parameter. This function seems to be largely independent of the implementation of `PreconditionerML`, which should be reflected in the software design in my opinion.

Next, I would think about integraging `Teuchos::ParameterList` into `MLData`.

Reply:
ok. I moved `Teuchos::ParameterList` to AMG_Data as I have "no" control over ML_Data in deal.II.